### PR TITLE
Allow loading non-contiguous spectrum number blocks in LoadISISNexus

### LINKED ISSUE #15482: Allow for loading new LOQ files with M3 and  M4 monitors.
LOQ will be using the M4 (beam stop monitor) soon for transmission measurements. This is currently not being handled in the SANS code. Also Stephen King has experienced problems loading the measuredd data set (becuase the monitor is not at  a spectrum index of 1 to 10)

The sample data set is LOQ95593 which has spectra ids as follows:
* 1-2: MON
* 3-17786: DET
* 17787-17788: MON
* 17789-17792: DET

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
Fixes #15482

**Context**

LOQ will add an M3 and M4 monitor soon. A first test run on Mantid failed. The reason for this because the way the detectors and monitors are wired up in the following manner:
|----Monitors--||---Detectors---||----Monitors---||----Detectors----|

This cannot be handled currently by the `LoadISISNexus` alogorithm. This issue tries to fix this.

**To test:**
- Please code review
- The unit tests cover several cases  (`LoadISISNexus` is indirectly being tested by a whole bunch of system tests. Since they will use it to load the reference data or test data). The cases of intersperesed monitors which should get handled by this issue cannot be tested directly at this point in time. The reason for this is that the required IDFs are not available or at least should not be in the IDF list of Mantid (not yet). A follow up issue which adds for these cases has been created here:

Please find the required test data here: \olympic\Babylon5\Scratch\Anton\Testing\Framework\PR_15562

**Loading workspace with monitors at the beginning**

The relevant workspace is: LOQ49886.nxs
To test use the instructions below.

**Loading workspace with monitors  at the end**

The relevant workspace is: MER18492.nxs
To test use the instructions below.

**Loading workspace with monitors which are interleaved with the detectors**

The relevant workspace is: LOQ95593.nxs
The relevant IDF is: LOQ_Definition_20160301.xml
You have to add this IDF to your repo your IDF folder. It is important that you remove it once you are done with testing this data set!

To test use the instructions below.

---

_Test instructions:_

Try each of the following instructions below with different settings for 
  `SpectrumMin`, `SpectrumMax`, `SpectrumList`,  `EntryNumber` 
- Press load and select the workspace
- Set:
  - LoadMonitors: Include
    *Confirm that:
  - Detectors and Monitors were loaded together
  - Spectrum numbers match your selection
- Press load and select the workspace
- Set:
  - LoadMonitors: Exclude
- Confirm that:
  - Only detectors are loaded
  - Spectrum numbers match your selection
- Press load and select the workspace
- Set:
  - LoadMonitors: Separate
- Confirm that:
  - Detectors and Monitors are loaded into separate workspaces
  - Spectrum numbers match your selection

[Release notes](https://github.com/mantidproject/mantid/blob/15482_implement_m4_functionality_for_LOQ/docs/source/release/v3.7.0/framework.rst)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
The other issue related to this (#5579) describes that INES files are a problem too. We should check those as well.

**Comment by martyngigg** [comment]
Actually we could use the old INES `.nxs` files as tests rather than the LOQ files.

**Comment by martyngigg** [comment]
I've checked `INS09161.nxs` and I correctly see the two monitors defined in the `INES` IDF marked correctly in the workspace, along with the right spectrum numbers.

**Comment by martyngigg** [comment]
As a general comment on the code I like that the helper classes have been factored out and tested. 

As they are solely to do with `LoadISISNexus2` I think they might be better off not being included in the `inc/MantidDataHandling` section and just be in the `src` folder as they are really implementation headers. Maybe consider creating a `LoadISISNexus` directory in `src` and moving the relevant files there?

**Comment by martyngigg** [comment]
@picatess Is this ready to test again? I notice it still has the "In Progress" label.

**Comment by picatess** [comment]
@martyngigg  I found an issue with the case when you include or exclude monitors and specify a range AND a list. Will fix this and let you know ASAP.

**Comment by picatess** [comment]
@martyngigg  The case of spectra list + spectrum list + monitors not at the start seems to be resolved. The latest installer can be found here: \\olympic\Babylon5\Scratch\Anton\Testing\Framework\PR_15562

**Comment by picatess** [comment]
@martyngigg : I just updated the branch. The `rst`-files for the relase notes were in conflict with a recently merged PR

File: Framework/DataHandling/test/LoadISISNexusTest.h [hunk 1]
```diff
@@ -769,6 +772,316 @@ class LoadISISNexusTest : public CxxTest::TestSuite {
...
+                 detIDtoWSIndexMap.count(monitorDetID) == 0);
+    }
+    for (const auto &neighborToCheck : neighborsToCheck) {
+      TSM_ASSERT("Should be in the detID2WSIndexMap.",
+                 detIDtoWSIndexMap.count(neighborToCheck) == 1);
+    }
+
+    // Check some of the data
+    double delta = 1e-6;
+    TS_ASSERT_DELTA(ws->readY(142)[0], 0.0, delta);
```
Comment (martyngigg) [inline Framework/DataHandling/test/LoadISISNexusTest.h]: The tests here and below are really nice and thorough for each of the cases regarding monitor data location. While we have the check that the monitor IDs have been loaded into the `detIDtoWSIndexMap` what I think we are missing is the check for the value of `getSpectrumNo()` and `getDetectorIDs()` for the spectra tested below.

I think we need this to be sure that the correct spectrum number/ID has been assigned to each histogram. The map only checks that they are there. I this needs to be checked for each of these test methods.

**Comment by martyngigg** [comment]
I'm happy with the code changes now. I will move on to functional testing.

**Comment by martyngigg** [comment]
Testing performed:
* Old-style file `LOQ49886.nxs` (all tested with all three `LoadMonitors` options):
 - no spectra specification
 - spectra min/max specification with spectra all monitors, all detectors and a mixture
 - spectra list specification with spectra all monitors, all detectors and a mixture

Issues encountered:
* Run `Load` with `LOQ49886.nxs`, set `SpectrumMin=1`, `SpectrumMax=1` & `LoadMonitors=Exclude`. I would expect this to throw an error as the request is the exclude monitors from loading but the only spectra selected are monitors. If however I just specify `SpectraList=1` then I get an error as expected (the error message is not great though).
* Run `Load` with `LOQ49886.nxs`, set `SpectraList=1,2,3` and `LoadMonitors=Separate`. It throws an error when it should succeed. If you specify `SpectrumMin=1`, `SpectrumMax=3` & `LoadMonitors=Exclude` then it succeeds as expected

File: Framework/DataHandling/src/LoadISISNexus2.cpp [hunk 2]
```diff
@@ -487,123 +507,190 @@ void LoadISISNexus2::checkOptionalProperties(
...
+
+    if (isSpectraListTooLarge) {
       std::string err =
-          "A spectra number in the spectra list exceeds maximal spectra ID:  " +
-          boost::lexical_cast<std::string>(m_loadBlockInfo.spectraID_max) +
-          " in the file ";
+          "The specified spectrum list contains a spectrum number which is "
+          "larger "
+          "than the largest loadable spectrum number for your selection of "
+          "excluded/included/seaparate monitors. The largest loadable "
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadISISNexus2.cpp]: There is a spelling mistake here seaparate -> separate

File: Framework/DataHandling/src/LoadISISNexus2.cpp [hunk 3]
```diff
@@ -487,123 +507,190 @@ void LoadISISNexus2::checkOptionalProperties(
...
+    if (isSpectraListTooSmall) {
       std::string err =
-          "A spectra number in the spectra list smaller then minimal spectra "
-          "ID:  " +
-          boost::lexical_cast<std::string>(m_loadBlockInfo.spectraID_min) +
-          " in the file";
+          "The specified spectrum list contains a spectrum number which is "
+          "smaller "
+          "than the smallest loadable spectrum number for your selection of "
+          "excluded/included/seaparate monitors. The smallest loadable "
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadISISNexus2.cpp]: There is a spelling mistake here seaparate -> separate

**Comment by martyngigg** [comment]
I'm happy this is working correctly now. If we could just correct the spelling mistake in the error message then I'll mark it ready to go.

**Comment by martyngigg** [comment]
I've also opened an issue for a problem with the detector table view that I encountered during testing -#15658. This was *not* caused by these changes.

**Comment by martyngigg** [comment]
:shipit:

**Pull request merged**